icicle_v3/backend/cpu/include/ntt_template.h

```diff
@@ -0,0 +1,264 @@
+#include "icicle/errors.h"
+#include "icicle/runtime.h"
+#include "icicle/utils/log.h"
+
+#include "icicle/fields/field_config.h"
+#include "icicle/utils/log.h"
+
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <cmath>
+#include <cstdint>
+
+using namespace field_config;
+using namespace icicle;
+namespace ntt_template {
+  template <typename S>
+  class CpuNttDomain
+  {
+    // TODO - coset, mixed-radix NTT
+    int max_size = 0;
+    int max_log_size = 0;
+    S* twiddles = nullptr;
+
+  public:
+    template <typename U>
+    friend eIcicleError
+    cpu_ntt_init_domain(const Device& device, const U& primitive_root, const NTTInitDomainConfig& config);
+
+    template <typename U>
+    friend eIcicleError generate_twiddles(const U& primitive_root, U* twiddles, int logn);
+
+    template <typename U>
+    friend eIcicleError cpu_ntt_release_domain(const Device& device);
+
+    template <typename U, typename E>
+    friend eIcicleError
+    cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    template <typename U, typename E>
+    friend eIcicleError
+    cpu_ntt(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    S* get_twiddles() { return twiddles; }
+  };
+
+  template <typename S>
+  static CpuNttDomain<S>* s_ntt_domain = nullptr;
+
+  template <typename S = scalar_t>
+  eIcicleError generate_twiddles(const S& primitive_root, S* twiddles, int n)
+  {
+    S omega = primitive_root;
+    twiddles[0] = S::one();
+    for (int i = 1; i <= n; i++) {
+      twiddles[i] = twiddles[i - 1] * omega;
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t>
+  eIcicleError cpu_ntt_init_domain(const Device& device, const S& primitive_root, const NTTInitDomainConfig& config)
+  {
+    // (1) check if need to refresh domain. TODO - reusing domain for different ntt sizes if possible
+    if (s_ntt_domain<S> != nullptr) { return eIcicleError::SUCCESS; }
+    // (2) build the domain
+    delete s_ntt_domain<S>;
+    s_ntt_domain<S> = new CpuNttDomain<S>();
+
+    bool found_logn = false;
+    S omega = primitive_root;
+    unsigned omegas_count = S::get_omegas_count();
+    for (int i = 0; i < omegas_count; i++) {
+      omega = S::sqr(omega);
+      if (!found_logn) {
+        ++s_ntt_domain<S>->max_log_size;
+        found_logn = omega == S::one();
+        if (found_logn) break;
+      }
+    }
+
+    s_ntt_domain<S>->max_size = (int)pow(2, s_ntt_domain<S>->max_log_size);
+    if (omega != S::one()) {
+      ICICLE_LOG_ERROR << "Primitive root provided to the InitDomain function is not in the subgroup";
+      return eIcicleError::INVALID_ARGUMENT;
+    }
+
+    // calculate twiddles
+    // Note: radix-2 INTT needs ONE in last element (in addition to first element), therefore have n+1 elements
+
+    s_ntt_domain<S>->twiddles = (S*)malloc((s_ntt_domain<S>->max_size + 1) * sizeof(S));
```

Comment:
I would avoid malloc/free and also new/delete and use smart pointers instead. In that case probably twiddles should be a unique_ptr.
Generally better avoid those since it avoids all the double frees and but of this kind and modern c++ has better solutions.

Even the Domain instance itself could be a value and not a pointer so that you don't need to call new and delete on it, if you want.

Reply:
fixed